File: examples/go/minimal-order-service/adapters/outbound/in_memory_repository.go
package outbound

import (
	"sync"
	"github.com/example/order-service/domain"
)

// InMemoryOrderRepository is an in-memory implementation of OrderRepository.
type InMemoryOrderRepository struct {
	orders map[string]*domain.Order
	mu     sync.RWMutex
}

// NewInMemoryOrderRepository creates a new InMemoryOrderRepository.
func NewInMemoryOrderRepository() *InMemoryOrderRepository {
	return &InMemoryOrderRepository{
		orders: make(map[string]*domain.Order),
	}
}

// Save saves an order.
func (r *InMemoryOrderRepository) Save(order *domain.Order) error {
	r.mu.Lock()
	defer r.mu.Unlock()
	r.orders[order.ID()] = order
	return nil
}

// FindByID finds an order by ID.
func (r *InMemoryOrderRepository) FindByID(id string) (*domain.Order, error) {
	r.mu.RLock()
	defer r.mu.RUnlock()
	order, exists := r.orders[id]
	if !exists {
		return nil, nil
	}
	return order, nil
}

File: examples/python/minimal-order-service/adapters/outbound/in_memory_repository.py
"""
Outbound adapter: In-memory implementation of OrderRepository.
Used for testing and simple scenarios.
"""
from typing import Optional
from application.ports.outbound.order_repository import OrderRepository
from domain.order import Order


class InMemoryOrderRepository(OrderRepository):
    """In-memory implementation of OrderRepository."""
    
    def __init__(self):
        """Initialize the repository."""
        self._orders: dict[str, Order] = {}
    
    def save(self, order: Order) -> None:
        """Save an order."""
        self._orders[order.id] = order
    
    def find_by_id(self, order_id: str) -> Optional[Order]:
        """Find an order by ID."""
        return self._orders.get(order_id)
